lib/gems/pending/util/mount/miq_swift_session.rb

```diff
@@ -0,0 +1,120 @@
+    super(log_settings)
+    raise "username and password are required values!" if @settings[:username].nil? || @settings[:password].nil?
+    _scheme, _userinfo, @host, @port, _registry, @mount_path, _opaque, query, _fragment = URI.split(URI.encode(@settings[:uri]))
+    query_params(query)
+    @username      = @settings[:username]
+    @password      = @settings[:password]
+    @container_name = @mount_path[0] == File::Separator ? @mount_path[1..-1] : @mount_path
+  end
+
+  def self.raw_disconnect(mnt_point)
+    return if mnt_point.nil?
+    FileUtils.rm_rf(mnt_point) if File.exist?(mnt_point)
+  end
+
+  def uri_to_local_path(remote_file)
+    # Strip off the leading "swift:/" from the URI"
+    File.join(@mnt_point, URI(remote_file).host, URI(remote_file).path)
+  end
+
+  def uri_to_object_path(remote_file)
+    # Strip off the leading "swift://" and the container name from the URI"
+    # Also remove the leading delimiter.
+    URI(remote_file).path[1..-1]
+  end
+
+  def add(local_file, uri)
+    begin
+      container = swift.directories.get(@container_name)
+    rescue Fog::Storage::OpenStack::NotFound
+      logger.debug("Swift container #{@container_name} does not exist.  Creating.")
+      begin
+        container = swift.directories.create(:key => @container_name)
+      rescue => err
+        disconnect
+        logger.error("Error creating Swift container #{@container_name}. #{err}")
+        msg = "Error creating Swift container #{@container_name}. #{err}"
+        raise err, msg, err.backtrace
+      end
+    rescue => err
+      disconnect
+      logger.error("Error getting Swift container #{@container_name}. #{err}")
+      msg = "Error getting Swift container #{@container_name}. #{err}"
+      raise err, msg, err.backtrace
+    end
+    #
+    # Get the remote path, and parse out the bucket name.
+    #
+    object_file_with_bucket = URI.split(URI.encode(uri))[5]
+    object_file = object_file_with_bucket.split(File::Separator)[2..- 1].join(File::Separator)
+    #
+    # write dump file to swift
+    #
+    logger.debug("Writing [#{local_file}] to Bucket [#{@container_name}] using object file name [#{object_file}]")
+    begin
+      container.files.create(:key => object_file, :body => File.open(local_file))
+    rescue Excon::Errors::Unauthorized => err
+      disconnect
+      logger.error("Access to Swift container #{@container_name} failed due to a bad username or password. #{err}")
+      msg = "Access to Swift container #{@container_name} failed due to a bad username or password. #{err}"
+      raise err, msg, err.backtrace
+    rescue => err
+      disconnect
+      logger.error("Error uploading #{local_file} to Swift container #{@container_name}. #{err}")
+      msg = "Error uploading #{local_file} to Swift container #{@container_name}. #{err}"
+      raise err, msg, err.backtrace
+    end
+  end
+
+  private
+
+  def swift
+    require 'manageiq/providers/openstack/legacy/openstack_handle'
```

Comment:
https://github.com/ManageIQ/manageiq-gems-pending/pull/371#issuecomment-419946719 was meant in response to this, but I moved it into my full review.

Answer to it can be found in https://github.com/ManageIQ/manageiq-gems-pending/pull/371#pullrequestreview-153815510

Reply:
For what it is worth, `OpenstackHandle::Handle` is basically a thin wrapper around `fog-openstack`, so we probably could just include that and use that directly.

I will try and see what is necessary in my branch to make the changes later if I have some time.

Note:  I still think the suggestion from above (move in to providers repos) is better long term, but agree that we are probably to close to the deadline to pull that off safely.

Reply:
So - FWIW there is similar code in https://github.com/ManageIQ/manageiq/pull/17967 for the same reason.  Yes its a wrapper around fog-openstack but I believe it to be a necessary one.  Short of pulling that code out into a repo/gem separate from the openstack provider, or moving this MountSession code into the Provider.